# Arrays vs Lists

In [1]:
import numpy as np

In [2]:
L = [1, 2, 3]

In [3]:
A = np.array([1, 2, 3])

In [4]:
for e in L:
    print(e)

1
2
3


In [5]:
for e in A:
    print(e)

1
2
3


Adding a new item to a list

In [6]:
L.append(4)

In [7]:
L

[1, 2, 3, 4]

In [8]:
A.append(4)

AttributeError: 'numpy.ndarray' object has no attribute 'append'

In [9]:
L + [ 5 ]

[1, 2, 3, 4, 5]

In [10]:
A + np.array([4])

array([5, 6, 7])

In [11]:
A + np.array([5, 6, 7])

array([ 6,  8, 10])

In [12]:
A + np.array([4, 5])

ValueError: operands could not be broadcast together with shapes (3,) (2,) 

## Scalar and Multiplication

In [15]:
2 * A

array([2, 4, 6])

In [17]:
2 * L

[1, 2, 3, 4, 1, 2, 3, 4]

In [19]:
L + L

[1, 2, 3, 4, 1, 2, 3, 4]

In [21]:
L2 = []
for e in L:
    L2.append(e + 3 )
L2

[4, 5, 6, 7]

In [23]:
L2 = [ e + 3 for e in L ]
L2

[4, 5, 6, 7]

In [24]:
L**2

TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'

In [25]:
L2 = []
for e in L:
    L2.append(e**2)

In [26]:
L2

[1, 4, 9, 16]

In [27]:
A**2

array([1, 4, 9])

In [28]:
np.sqrt(A)

array([1.        , 1.41421356, 1.73205081])

In [29]:
np.log(A)

array([0.        , 0.69314718, 1.09861229])

In [30]:
np.exp(A)

array([ 2.71828183,  7.3890561 , 20.08553692])

In [31]:
np.tanh(A)

array([0.76159416, 0.96402758, 0.99505475])

# The Dot Product

$$ a \cdot b = a^T b = \sum_{d=1}^D a_d b_d $$

In [32]:
a = np.array([1, 2])
b = np.array([3, 4])

In [33]:
dot = 0
for e, f in zip(a, b):
    dot += e * f
dot

11

In [34]:
dot = 0
for i in range(len(a)):
    dot += a[i] * b[i]
dot

11

In [35]:
a * b

array([3, 8])

In [36]:
np.sum(a * b)

11

In [37]:
(a * b).sum()

11

In [38]:
np.dot(a, b)

11

In [39]:
a.dot(b)

11

In [40]:
a @ b

11

$$ a^T b = \left\lVert a \right\rVert \left\lVert b \right\rVert \cos \theta_{ab}$$

$$ \cos \theta_{ab} = \frac{a^T b}{\left\lVert a \right\rVert \left\lVert b \right\rVert} $$

$$ \left\lVert a \right\rVert = \sqrt{\sum_{d=1}^D a_d^2 } $$

In [41]:
amag = np.sqrt(( a * a ).sum())

In [42]:
amag

2.23606797749979

In [43]:
np.linalg.norm(a)

2.23606797749979

In [44]:
cosangle = a.dot(b) / (np.linalg.norm(a) * np.linalg.norm(b))
cosangle

0.9838699100999074

In [45]:
angle = np.arccos(cosangle)
angle

0.17985349979247847

# SpeedTest

In [46]:
## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.random.randn(100)
b = np.random.randn(100)
T = 100000

def slow_dot_product(a, b):
  result = 0
  for e, f in zip(a, b):
    result += e*f
  return result

t0 = datetime.now()
for t in range(T):
  slow_dot_product(a, b)
dt1 = datetime.now() - t0

t0 = datetime.now()
for t in range(T):
  a.dot(b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())

dt1 / dt2: 39.61785592875261


# Matrices

In [47]:
L = [[1,2],[3,4]]
L

[[1, 2], [3, 4]]

In [48]:
L[0]

[1, 2]

In [49]:
L[0][1]

2

In [50]:
A = np.array([[1,2],[3,4]])

In [51]:
A

array([[1, 2],
       [3, 4]])

In [52]:
A[0][1]

2

In [53]:
A[0,1]

2

In [54]:
A[:,0]

array([1, 3])

In [55]:
A.T

array([[1, 3],
       [2, 4]])

In [56]:
np.exp(A)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [57]:
np.exp(L)

array([[ 2.71828183,  7.3890561 ],
       [20.08553692, 54.59815003]])

In [58]:
B = np.array([[1,2,3],[4,5,6]])
B

array([[1, 2, 3],
       [4, 5, 6]])

In [59]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [60]:
A

array([[1, 2],
       [3, 4]])

In [61]:
B

array([[1, 2, 3],
       [4, 5, 6]])

In [62]:
A.dot(B)

array([[ 9, 12, 15],
       [19, 26, 33]])

In [63]:
A.dot(B.T)

ValueError: shapes (2,2) and (3,2) not aligned: 2 (dim 1) != 3 (dim 0)

In [64]:
np.linalg.det(A)

-2.0000000000000004

In [65]:
np.linalg.inv(A)

array([[-2. ,  1. ],
       [ 1.5, -0.5]])

In [66]:
np.linalg.inv(A).dot(A)

array([[1.00000000e+00, 0.00000000e+00],
       [2.22044605e-16, 1.00000000e+00]])

In [67]:
np.trace(A)

5

In [68]:
A

array([[1, 2],
       [3, 4]])

In [69]:
np.diag(A)

array([1, 4])

In [70]:
np.diag([1,4])

array([[1, 0],
       [0, 4]])

In [71]:
np.linalg.eig(A)

(array([-0.37228132,  5.37228132]),
 array([[-0.82456484, -0.41597356],
        [ 0.56576746, -0.90937671]]))

In [72]:
Lam, V = np.linalg.eig(A)

In [73]:
V[:,0] * Lam[0] == A @ V[:,0]

array([ True, False])

In [74]:
V[:,0] * Lam[0], A @ V[:,0]

(array([ 0.30697009, -0.21062466]), array([ 0.30697009, -0.21062466]))

In [75]:
np.allclose(V[:,0] * Lam[0], A @ V[:,0])

True

In [76]:
np.allclose(V @ np.diag(Lam), A @ V)

True

# Linear Systems: Example problem

The admission fee at a small fair is \$1.50 for children and \$4 for adults. On a certain day, 2200 people enter the fair, and $5050 is collected. How many children and adults attended?

2 equations, 2 unknowns

$ x1 + x2 = 2200 $   
$ 1.5x1 + 4x2 = 5050 $

## Linear System Matrix form

$$ x= \begin{pmatrix} x1 \\ x2 \end{pmatrix}, A = \begin{pmatrix} 1 & 1 \\  1.5 & 4 \end{pmatrix}, B = \begin{pmatrix} 2200 \\ 5050 \end{pmatrix} $$   

$$ Ax = b \iff x = A^{-1} b$$

In [77]:
A = np.array([[1, 1], [1.5, 4]])
b = np.array([2200, 5050])

In [78]:
#Solve a linear matrix equation, or system of linear scalar equations.
np.linalg.solve(A, b)

array([1500.,  700.])

In [79]:
# don't do this
np.linalg.inv(A).dot(b)

array([1500.,  700.])

# Generating Data

In [80]:
# Generating a matrix full of zeros
np.zeros((2, 3))

array([[0., 0., 0.],
       [0., 0., 0.]])

In [81]:
# Generating a matrix full of ones
np.ones((2, 3))

array([[1., 1., 1.],
       [1., 1., 1.]])

In [82]:
# What if we want a Matrix full of a number, we just simply multiply the number by one
6 * np.ones((2, 3))

array([[6., 6., 6.],
       [6., 6., 6.]])

In [84]:
# Generating an identity matrix with eye
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [85]:
# Generate a random number
np.random.random()

0.5007426302455609

In [86]:
# Generate a matrix with random numbers
np.random.random((2,3))

array([[0.39550479, 0.21156248, 0.44772775],
       [0.38873674, 0.77969368, 0.61469348]])

In [87]:
# Generate random numbers from gaussian distribution
np.random.randn(2,3)

array([[-0.88494177,  1.08469266,  1.02706734],
       [ 1.65373923,  2.54643586, -0.79310988]])

In [88]:
# Generate a large matrix
R = np.random.randn(10000)

In [89]:
# Calculate the arithmetic mean value (media in spanish)
R.mean()

-0.022097202329907337

In [90]:
np.mean(R)

-0.022097202329907337

In [91]:
# Calculate the variance
R.var()

1.0093593600831792

In [92]:
# Calculate standard deviation (Desviación estándar in spanish)
R.std()

1.0046687812822588

In [93]:
# Generate a 10,000 x 3 matrix of random numbers
# 10,000 rows x 3 columns
R = np.random.randn(10000,3)

In [94]:
# Calculates the mean of each column
R.mean(axis=0)

array([ 0.00157299, -0.00616391,  0.00147782])

In [96]:
R.mean(axis=1).shape

(10000,)

In [98]:
np.cov(R).shape

(10000, 10000)

In [99]:
 np.cov(R.T)

array([[ 0.99860316, -0.01074068,  0.01941845],
       [-0.01074068,  1.00714591, -0.00108781],
       [ 0.01941845, -0.00108781,  1.02666367]])

In [100]:
np.random.randint(0, 10, size=(3,3))

array([[3, 7, 2],
       [3, 5, 8],
       [5, 5, 9]])

In [101]:
np.random.choice(10, size=(3,3))

array([[9, 1, 5],
       [0, 1, 4],
       [5, 2, 2]])

# Numpy Exercise

Create a speed test for matrix multiplication with Numpy Arrays and Lists

In [104]:
# Matrices
# A = 3 x 2
# B = 2 x 3

## speed comparison ##
import numpy as np
from datetime import datetime

# note: you can also use %timeit

a = np.array([[3,1],[2,6],[5,7]])

b = np.array([[4,6,8],[1,2,6]])

sum = np.zeros((2,2))

# List multiplication matrix
t0 = datetime.now()

for h in range(2):
    for i in range(2):
        for j in range(2):
            sum[h][i] += a[h][j] * b[j][i]

dt1 = datetime.now() - t0

# Numpy Array multiplication
t0 = datetime.now()
a.dot(b)
dt2 = datetime.now() - t0

print("dt1 / dt2:", dt1.total_seconds() / dt2.total_seconds())


dt1 / dt2: 0.3355648535564853
